lib/nerves_runtime/configfs.ex

```diff
@@ -0,0 +1,153 @@
+defmodule Nerves.Runtime.ConfigFS do
+  use GenServer
+  alias Nerves.Runtime
+  require Logger
+
+  def start_link() do
+    GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  end
+
+  def init([]) do
+    # configfs_home="/sys/kernel/config"
+    # mount none $configfs_home -t configfs
+    case Runtime.cmd("mount", [], :return) do
+      {str, 0} -> String.contains?(str, "none on /sys/kernel/config type configfs")
+      {err, _} -> raise "Failed to check mounted partitions: #{inspect(err)}"
+    end
+    |> unless do
+      Logger.info("Going to mount ConfigFS.")
+      {_, 0} = Runtime.cmd("mount", ["none", "/sys/kernel/config", "-t", "configfs"], :return)
+    end
+
+    apply_map_to_configfs(gadget_config())
+    g = "/sys/kernel/config/usb_gadget/g"
+    apply_link(Path.join(g, "functions/ecm.usb0"), Path.join(g, "configs/c.1"))
+    apply_link(Path.join(g, "functions/acm.usb0"), Path.join(g, "configs/c.1"))
+
+    apply_link(Path.join(g, "functions/rndis.usb0"), Path.join(g, "configs/c.2"))
+    # apply_link(Path.join(g, "functions/acm.usb0"), Path.join(g, "configs/c.2"))
+
+    apply_link(Path.join(g, "configs/c.2"), Path.join(g, "os_desc"))
+
+    {_, 0} = Runtime.cmd("sh", ["-c", "ls /sys/class/udc > #{Path.join(g, "UDC")}"], :return)
+    {:ok, %{}}
+  end
+
+  def apply_map_to_configfs(map) do
+    mani = build_manifest(map)
+
+    Enum.each(Enum.reverse(mani.folders), fn folder ->
+      mkdir_p(Path.join("/sys/kernel/config/", folder))
+    end)
+
+    Enum.each(Enum.reverse(mani.files), fn {file, value} ->
+      write(Path.join("/sys/kernel/config/", file), value)
+    end)
+  end
+
+  defp apply_link(patha, pathb) do
+    ln_s(patha, pathb)
+  end
+
+  defp mkdir_p(dir) do
+    {_, 0} = Runtime.cmd("mkdir", ["-p", dir], :return)
+  end
+
+  defp ln_s(patha, pathb) do
+    {_, 0} = Runtime.cmd("ln", ["-s", patha, pathb], :return)
+  end
+
+  defp write(path, value) do
+    {_, 0} = Runtime.cmd("sh", ["-c", "echo #{value} > #{path}"], :return)
```

Comment:
`File.write!` doesn't work on these files? Could you test? If not, OTP 21 can't come fast enough.